aws-sdk-v3 upgrade and 1.0.0 release

petermyers opened the pull request:
Changes tested by publishing v1.0.0 to locally running Verdaccio and using [these test pipelines](https://github.com/petermyers/aws-lambda-stream-examples/tree/main/example-connector-test) to verify connectors.

uuid util added to enable mocking since the newer version of the uuid library exports modules in a way that can't be mocked with Sinon.

log util added since the logger interface passable to aws client constructors became slightly more complex and I didn't want to duplicate that logic in every connector.

jgilbert01 commented on src/connectors/kinesis.js:
```diff
@@ -17,13 +16,12 @@ class Connector {
+        requestTimeout: timeout,
+        connectionTimeout: timeout,
+      }),
+      maxAttempts: 11,
```
are there no longer retry delay options?

petermyers replied:
I found a way to do this, it requires supplying the whole function in a `ConfiguredRetryStrategy`. Added this in to the connectors that previously had retry delay options.

jgilbert01 commented on src/connectors/s3.js:
```diff
@@ -60,7 +58,8 @@ class Connector {
 
-    return this.bucket.getObject(params).createReadStream();
+    return this._sendCommand(new GetObjectCommand(params))
+      .then((response) => Readable.from(response.Body));
```
i think this is changing the functionality because we have to get the full body first

jgilbert01 replied:
transformToWebStream ?

jgilbert01 replied:
https://github.com/smithy-lang/smithy-typescript/blob/280ef3a90205bfaa208fe9ddfc088a80341c5032/packages/util-stream/src/sdk-stream-mixin.ts#L32

petermyers replied:
Does creating a `Readable` cause the incoming body stream to be read? I thought it would create the stream based on the async iterable in an initially paused state.

I did tried using `transformToWebStream`, which theoretically returns a `ReadableStream` but Highland refuses to wrap it and throws an error: `Object was not a stream, promise, iterator or iterable: object`.

I'll keep playing with it though.

jgilbert01 commented on src/connectors/s3.js:
```diff
@@ -49,9 +48,8 @@ class Connector {
-      .tap(this.debug)
-      .tapCatch(this.debug);
+    return this._sendCommand(new GetObjectCommand(params))
+      .then(async (response) => ({ ...response, Body: Buffer.concat(await response.Body.toArray()).toString() }));
```
is this different than response.Body.transformToString()

petermyers replied:
It was not functionally different, updated to use `transformToString`.

jgilbert01 commented:
to maximize v3 modularity, i'm thinking we should remove the various index files and move to using module-based imports.

for example:
`import Connector from 'aws-lambda-stream/connectors/dynamodb';`

instead of: 
`import { DynamodbConnector }  from 'aws-lambda-stream';`

because the index files force the need to include all the dependencies.

we could leave the index files for backward compatibility but deprecate them and stop adding anything to them

thoughts?

petermyers commented:
> to maximize v3 modularity, i'm thinking we should remove the various index files and move to using module-based imports.
> 
> for example: `import Connector from 'aws-lambda-stream/connectors/dynamodb';`
> 
> instead of: `import { DynamodbConnector } from 'aws-lambda-stream';`
> 
> because the index files force the need to include all the dependencies.
> 
> we could leave the index files for backward compatibility but deprecate them and stop adding anything to them
> 
> thoughts?

Makes sense to me, I can make those changes. I think what I'll do in the vein of compatibility is:
- Include any new exports I added to the index files for this release
- Add deprecation comments to the index files
- Enable module based imports

Then moving forward we can stop adding anything new to them.

petermyers commented:
> > to maximize v3 modularity, i'm thinking we should remove the various index files and move to using module-based imports.
> > for example: `import Connector from 'aws-lambda-stream/connectors/dynamodb';`
> > instead of: `import { DynamodbConnector } from 'aws-lambda-stream';`
> > because the index files force the need to include all the dependencies.
> > we could leave the index files for backward compatibility but deprecate them and stop adding anything to them
> > thoughts?
> 
> Makes sense to me, I can make those changes. I think what I'll do in the vein of compatibility is:
> 
> * Include any new exports I added to the index files for this release
> * Add deprecation comments to the index files
> * Enable module based imports
> 
> Then moving forward we can stop adding anything new to them.

You can take a look at [here](https://github.com/petermyers/aws-lambda-stream-examples/blob/main/example-connector-test/src/flavors/sns-connector.js) for an example of what usage looks like with modular exports. I updated the whole `example-connector-test` to use the modular exports, so feel free to take a look through that directory in my test repo.

jgilbert01 commented:
- this major release is also an opportunity to consider breaking the utils folder into a sink folder and a utils
- essentially "through" utils that invoke connectors r sinks

jgilbert01 commented:
did not mean to close

petermyers commented:
> * this major release is also an opportunity to consider breaking the utils folder into a sink folder and a utils
> * essentially "through" utils that invoke connectors r sinks

I had given that some thought and wanted to pick your brain. I'm generally not a big fan of general 'utils' unless they truly contain utilities like uuid generation, defaults, helpful small utilities to reduce code duplication.

The utils directory as it is now in my mind actually contains 3 things, sinks, through's, and actual utilities. I'm not sold on the name 'throughs', maybe 'stream-utils'? My idea is basically sinks, utilities that operate on the stream (eg. encryption, batching, faults), and basic utilities (eg. retry, opt, print)

My initial thought would be to split the files apart into the following dirs:
- sinks
  - cloudwatch
  - dynamodb
  - eventbridge
  - fetch
  - firehose
  - kinesis
  - lambda
  - s3
  - secretsmgr
  - sns
  - sqs
- through
  - batch
  - claimcheck
  - encryption
  - faults
  - handler
  - ratelimit
  - split
- utils
  - compression
  - log
  - opt
  - print
  - retry
  - tags
  - time
  - uow
  - uuid

Thoughts?

jgilbert01 commented:
as a first cut to reorganize i created sinks and queries folders and moved utils accordingly

jgilbert01 commented:
i also removed everything that was deprecated

jgilbert01 commented on src/queries/dynamodb.js:
```diff
@@ -120,17 +20,27 @@ export const batchGetDynamoDB = ({
-        }, {}),
-      }))
+    const req = JSON.stringify(uow[batchGetRequestField]);
+    const cached = memoryCache.get(req);
```
lots of room to improve on batch get caching

petermyers replied:
Definitely. I started here because it mimics how queryAll caching is being done atm. Have some thoughts here, we can discuss.

jgilbert01 replied:
this is a good safe starting position so i'll leave it

jgilbert01 commented on test/unit/utils/compression.test.js:
```diff
@@ -19,11 +19,13 @@ describe('utils/compression.js', () => {
     expect(event.length).to.equal(3317);
-    expect(compressed.length).to.equal(756);
+    // TODO - In Github actions this is 760, but in my OSX environment, it's 756.
+    // expect(compressed.length).to.equal(756);
```
- i get 760 locally for v3 and 756 locally for v2
- could it be a different in the kms-ee mock responses?
